src/ansible_navigator/actions/exec.py

```diff
@@ -0,0 +1,116 @@
+"""Run the :exec subcommand."""
+import os
+import shlex
+
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from . import _actions as actions
+from ..app import App
+from ..configuration_subsystem import ApplicationConfiguration
+from ..configuration_subsystem.definitions import Constants
+from ..runner import Command
+
+GeneratedCommand = Tuple[str, Optional[List[str]]]
+
+
+def _generate_command(exec_command: str, exec_shell: bool) -> GeneratedCommand:
+    """Generate the command and args.
+
+    :param exec_command: The command to run
+    :param exec_shell: Should the command be wrapped in a shell
+    :returns: The command and any pass through arguments
+    """
+    pass_through_args = None
+    if exec_shell and exec_command:
+        command = "/bin/bash"
+        pass_through_args = ["-c", exec_command]
+    else:
+        parts = shlex.split(exec_command)
+        command = parts[0]
+        if len(parts) > 1:
+            pass_through_args = parts[1:]
+    return (command, pass_through_args)
+
+
+@actions.register
+class Action(App):
+    """Run the :exec subcommand."""
+
+    # pylint: disable=too-few-public-methods
+
+    KEGEX = "^e(?:xec)?$"
+
+    def __init__(self, args: ApplicationConfiguration):
+        """Initialize the action.
+
+        :param args: The current application configuration.
+        """
+        super().__init__(args=args, logger_name=__name__, name="exec")
+
+    def run_stdout(self) -> Union[None, int]:
+        """Run in mode stdout.
+
+        :returns: The return code or None
+        """
+        self._logger.debug("exec requested in stdout mode")
+        response = self._run_runner()
+        if response:
+            _, _, ret_code = response
+            return ret_code
+        return None
+
+    def _run_runner(self) -> Optional[Tuple]:
+        # pylint: disable=too-many-branches
+        # pylint: disable=too-many-statements
+        """Spin up runner.
+
+        :return: The stdout, stderr and return code from runner
+        """
+        if isinstance(self._args.set_environment_variable, dict):
+            envvars_to_set = self._args.set_environment_variable.copy()
+        elif isinstance(self._args.set_environment_variable, Constants):
+            envvars_to_set = {}
+        else:
+            log_message = (
+                "The setting 'set_environment_variable' was neither a dictionary"
+                + " or Constants, please raise an issue. No environment variables will be set."
```

Comment:
Prefer parse-time concatenation over runtime. With this, the interpreter will treat both lines as a single string literal:
```suggestion
                " or Constants, please raise an issue. No environment variables will be set."
```
But with a `+`, it would parse two literals and only concatenate them when the control flow reaches this place in runtime, which is slightly less performant.

Reply:
Interesting, don't we introduce implicit string concatenation if we do that while it is explicit with the +?

Reply:
Yep. Exactly. Except with "implicit" concatenation, there's no actual concatenation in runtime, the interpreter would "see" the whole thing as if you typed in everything as one big literal. Are you hitting any linter complaints? I think there was a WPS rule that I'd always disable because it doesn't allow implicit concat. But I think having it implicit is more common and reasonable.

Reply:
https://github.com/ansible/ansible-navigator/pull/740